src/main/java/io/supertokens/signingkeys/SigningKeys.java

```diff
@@ -0,0 +1,275 @@
+        }
+
+        return res;
+    }
+
+    public List<JWTSigningKeyInfo> getStaticKeys() throws StorageQueryException, StorageTransactionLogicException {
+        if (this.staticKeys == null) {
+            this.staticKeys = JWTSigningKey.getInstance(main).getAllSigningKeys();
+        }
+
+        return this.staticKeys;
+    }
+    public KeyInfo getLatestIssuedDynamicKey() throws StorageQueryException, StorageTransactionLogicException {
+        return getDynamicKeys().get(0);
+    }
+
+    public synchronized long getDynamicSigningKeyExpiryTime() throws StorageQueryException, StorageTransactionLogicException {
+        long createdAtTime = getLatestIssuedDynamicKey().createdAtTime;
+        return createdAtTime + Config.getConfig(main).getAccessTokenDynamicSigningKeyUpdateInterval();
+    }
+
+    public synchronized void removeKeyFromMemoryIfItHasNotChanged(List<JWTSigningKeyInfo> oldKeyInfo)
+            throws StorageQueryException, StorageTransactionLogicException {
+        // we cannot use read write locks for keyInfo because in getKey, we would
+        // have to upgrade from the readLock to a
+        // writeLock - which is not possible:
+        // https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/ReentrantReadWriteLock.html
+
+        if (this.dynamicKeys != null && this.staticKeys != null &&
+                // we want to refresh if all keys we are storing were known before
+            this.getAllKeys().stream().allMatch(storedKey -> oldKeyInfo.stream().anyMatch(oldKey -> Objects.equals(oldKey.keyId, storedKey.keyId)))) {
+            // key has not changed since we previously tried to use it... So we can make it null.
+            // otherwise we might end up making this null unnecessarily.
+
+            ProcessState.getInstance(this.main)
+                    .addState(ProcessState.PROCESS_STATE.SETTING_ACCESS_TOKEN_SIGNING_KEY_TO_NULL, null);
+            this.staticKeys = null;
+            this.dynamicKeys = null;
+        }
+    }
+
+    /**
+     * Used to return public keys that a JWT verifier will use. Note returns an empty array if there are no keys in
+     * storage.
+     *
+     * @return JSON array containing the JWKs
+     * @throws StorageQueryException            If there is an error interacting with the database
+     * @throws StorageTransactionLogicException If there is an error interacting with the database
+     * @throws NoSuchAlgorithmException         If there is an error when using Java's cryptography packages
+     * @throws InvalidKeySpecException          If there is an error when using Java's cryptography packages
+     */
+    public List<JsonObject> getJWKS() throws StorageQueryException, StorageTransactionLogicException,
+            NoSuchAlgorithmException, InvalidKeySpecException {
+        List<JsonObject> jwks = new ArrayList<>();
+
+        // Retrieve all static keys in storage
+        List<JWTSigningKeyInfo> keys = this.getAllKeys();
```

Comment:
getAllKeys should return all the keys (static or not right?) So the comment is confusing. Also, getllKeys doesn't guaranteed hit the storage (cause the keys are cached in the dynamicKeys and staticKeys variable - is that OK?

Maybe change the comment to be more clear

Reply:
I think it's OK, but we could make it hit as well. It's OK if either of 2 assumptions holds:
- This is used through a BE SDK since that will call a verification endpoint that will refresh the cache if necessary
- There is only a single core connecting to a DB
Static keys are OK anyway since they never expire, and we generate all (1, currently) keys during startup.